streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java

```diff
@@ -157,7 +158,8 @@ int commitTasksAndMaybeUpdateCommittableOffsets(final Collection<Task> tasksToCo
             if (task.commitNeeded()) {
                 task.clearTaskTimeout();
                 ++committed;
-                task.postCommit(false);
+                // under EOS, we need to enforce a checkpoint if our transaction buffers will exceeded their capacity
```

Comment:
Shouldn't our transaction buffers be empty at this point?
We committed the transaction which means we should have also committed the state store.

Reply:
Actually, at this point we've only committed our Kafka transaction. The StateStore has not been committed/flushed, so our transaction buffers will still be full.

`task.postCommit` is the process that "checkpoints" state stores. We need to enforce a checkpoint when the transaction buffer is full, because we only flush the StateStore when it is checkpointed. See `AbstractTask` line `99`.

Reply:
There aren't actually any existing tests at all for `TaskExecutor#commitTasksAndMaybeUpdateCommittableOffsets`. I can still add some tests specifically for these changes, but I suspect that a lot more of the behaviour here is relying on integration tests.